File: src/MagnusAddon/language_services/conjugator.py
from __future__ import annotations

from typing import TYPE_CHECKING

import mylog

if TYPE_CHECKING:
    from note.vocabulary.vocabnote import VocabNote

_ichidan_endings = ["", "ろ", "な"]

_godan_ru_endings = ["り", "ら", "れ", "っ"]
_godan_ru_or_ichidan_endings = _godan_ru_endings + _ichidan_endings

godan_potential_verb_ending_to_dictionary_form_endings: dict[str, str] = {"える": "う", "ける": "く", "げる": "ぐ", "せる": "す", "てる": "つ", "ねる": "ぬ", "べる": "ぶ", "める": "む", "れる": "る"}

e_stem_hiragana: set[str] = {"え", "け", "げ", "せ", "て", "ね", "べ", "め", "れ"}
e_stem_katakana: set[str] = {"エ", "ケ", "ゲ", "セ", "テ", "ネ", "ベ", "メ", "レ"}
e_stem_characters: set[str] = e_stem_hiragana | e_stem_katakana

a_stem_hiragana: set[str] = {"わ", "か", "が", "さ", "た", "な", "ば", "ま", "ら", }
a_stem_katakana: set[str] = {"ワ", "カ", "ガ", "サ", "タ", "ナ", "バ", "マ", "ラ"}
a_stem_characters: set[str] = a_stem_hiragana | a_stem_katakana

# o_row_hiragana: set[str] = {"を", "こ", "ご", "そ", "と", "の", "ぼ", "も", "ろ"}
# o_row_katakana: set[str] = {"オ", "コ", "ゴ", "ソ", "ト", "ノ", "ボ", "モ", "ロ"}
# passive_form_stem_characters: set[str] = a_stem_characters - {"な", "ナ"} # in practice な only brings false positives when searching for passive since passive ぬ verb usages are non-existent

_i_stem_index = 0
_a_stem_index = 1
_e_stem_index = 2
_te_stem_index = 3
_1_character_mappings: dict[str, list[str]] = {"う": ["い", "わ", "え", "っ"],
                                               "く": ["き", "か", "け", "い"],
                                               "ぐ": ["ぎ", "が", "げ", "い"],
                                               "す": ["し", "さ", "せ", "し"],
                                               "つ": ["ち", "た", "て", "っ"],
                                               "ぬ": ["に", "な", "ね", "ん"],
                                               "ぶ": ["び", "ば", "べ", "ん"],
                                               "む": ["み", "ま", "め", "ん"],
                                               "る": _godan_ru_endings,
                                               "い": ["く", "け", "か"]}

_2_character_mappings: dict[str, list[str]] = {"する": ["し", "さ", "すれ", "し", "せ"],
                                               "くる": ["き", "こ", "くれ", "き"],
                                               "いく": ["いき", "いか", "いけ", "いっ", "いこ"],
                                               "行く": ["行き", "行か", "行け", "行っ", "行こ"],
                                               "ます": ["まし", "ませ"],
                                               "いい": ["よく", "よけ", "よか", "よかっ"]}

_masu_forms_by_index = ["まし", "ませ", "まし", "まし"]  # not to sure about these. To say the least....

_aru_verbs: set[str] = {"なさる", "くださる", "おっしゃる", "ござる", "らっしゃる", "下さる", "為さる"}

_aru_mappings: dict[str, list[str]] = {"さる": ["さい", "さら", "され", "さっ"],
                                       "ざる": ["ざい", "ざら", "ざれ", "ざっ"],
                                       "ゃる": ["ゃい", "ゃら", "れば", "ゃっ"]}

def construct_root_verb_for_possibly_potential_godan_verb_dictionary_form(potential_verb_form: str) -> str:
    return potential_verb_form[:-2] + godan_potential_verb_ending_to_dictionary_form_endings[potential_verb_form[-2:]]

def _is_aru_verb(word: str) -> bool:
    return any(aru_ending for aru_ending in _aru_verbs if word.endswith(aru_ending))

def get_word_stems(word: str, is_ichidan_verb: bool = False, is_godan: bool = False) -> list[str]:
    try:
        if _is_aru_verb(word):
            return [word[:-2] + end for end in _aru_mappings[word[-2:]]]
        if is_ichidan_verb:
            return [word[:-1] + end for end in _ichidan_endings]
        if is_godan:
            return [word[:-1] + end for end in _1_character_mappings[word[-1]]]
        if word[-2:] in _2_character_mappings:
            return [word[:-2] + end for end in _2_character_mappings[word[-2:]]]
        if word[-1] in _1_character_mappings:
            if word[-1] == "る":
                return [word[:-1] + end for end in _godan_ru_or_ichidan_endings]
            return [word[:-1] + end for end in _1_character_mappings[word[-1]]]
    except KeyError:
        mylog.warning(f"get_word_stems failed to handle {word}, returning empty list ")
    return [word]

def _get_stem(word: str, stem_index: int, is_ichidan_verb: bool = False, is_godan: bool = False) -> str:
    try:
        if _is_aru_verb(word):
            return word[:-2] + _aru_mappings[word[-2:]][stem_index]
        if is_ichidan_verb:
            return word[:-1]
        if is_godan:
            return word[:-1] + _1_character_mappings[word[-1]][stem_index]
        if word[-2:] == "ます":
            return word[:-2] + _masu_forms_by_index[stem_index]
        if word[-2:] in _2_character_mappings:
            return word[:-2] + _2_character_mappings[word[-2:]][stem_index]
        if word[-1] in _1_character_mappings:
            if word[-1] != "る":
                return word[:-1] + _1_character_mappings[word[-1]][stem_index]
            return word[:-1] + _1_character_mappings[word[-1]][stem_index]
    except KeyError:
        mylog.warning(f"_get_stem failed to handle {word}, returning empty list ")
    return word

def get_i_stem(word: str, is_ichidan: bool = False, is_godan: bool = False) -> str:
    return _get_stem(word, _i_stem_index, is_ichidan, is_godan)

def get_a_stem(word: str, is_ichidan: bool = False, is_godan: bool = False) -> str:
    return _get_stem(word, _a_stem_index, is_ichidan, is_godan)

def get_e_stem(word: str, is_ichidan: bool = False, is_godan: bool = False) -> str:
    return _get_stem(word, _e_stem_index, is_ichidan, is_godan)

def get_te_stem(word: str, is_ichidan: bool = False, is_godan: bool = False) -> str:
    return _get_stem(word, _te_stem_index, is_ichidan, is_godan)

def get_i_stem_vocab(vocab: VocabNote, form: str = "") -> str:
    return get_i_stem(form if form else vocab.get_question(), vocab.parts_of_speech.is_ichidan(), vocab.parts_of_speech.is_godan())

def get_e_stem_vocab(vocab: VocabNote, form: str = "") -> str:
    return get_e_stem(form if form else vocab.get_question(), vocab.parts_of_speech.is_ichidan(), vocab.parts_of_speech.is_godan())

def get_a_stem_vocab(vocab: VocabNote, form: str = "") -> str:
    return get_a_stem(form if form else vocab.get_question(), vocab.parts_of_speech.is_ichidan(), vocab.parts_of_speech.is_godan())

def get_te_stem_vocab(vocab: VocabNote, form: str = "") -> str:
    return get_te_stem(form if form else vocab.get_question(), vocab.parts_of_speech.is_ichidan(), vocab.parts_of_speech.is_godan())

File: src/src_dotnet/JAStudio.Core/LanguageServices/Conjugator.cs
using JAStudio.Core.Note;
using System.Collections.Generic;
using System.Linq;

namespace JAStudio.Core.LanguageServices;

// ReSharper disable UnusedMember.Global
// ReSharper disable InconsistentNaming
public static class Conjugator
{
   static readonly List<string> IchidanEndings = ["", "ろ", "な"];

   static readonly List<string> GodanRuEndings = ["り", "ら", "れ", "っ"];
   static readonly List<string> GodanRuOrIchidanEndings;

    public static readonly Dictionary<string, string> GodanPotentialVerbEndingToDictionaryFormEndings = new()
    {
        { "える", "う" },
        { "ける", "く" },
        { "げる", "ぐ" },
        { "せる", "す" },
        { "てる", "つ" },
        { "ねる", "ぬ" },
        { "べる", "ぶ" },
        { "める", "む" },
        { "れる", "る" }
    };

    public static readonly HashSet<string> GodanImperativeVerbEndings = ["え", "け", "げ", "せ", "て", "ね", "べ", "め", "れ"];

    const int IStemIndex = 0;
    const int AStemIndex = 1;
    const int EStemIndex = 2;
    const int TeStemIndex = 3;

    static readonly Dictionary<string, List<string>> OneCharacterMappings = new()
                                                                            {
                                                                               { "う", ["い", "わ", "え", "っ"] },
                                                                               { "く", ["き", "か", "け", "い"] },
                                                                               { "ぐ", ["ぎ", "が", "げ", "い"] },
                                                                               { "す", ["し", "さ", "せ", "し"] },
                                                                               { "つ", ["ち", "た", "て", "っ"] },
                                                                               { "ぬ", ["に", "な", "ね", "ん"] },
                                                                               { "ぶ", ["び", "ば", "べ", "ん"] },
                                                                               { "む", ["み", "ま", "め", "ん"] },
                                                                               { "る", GodanRuEndings },
                                                                               { "い", ["く", "け", "か"] } // I adjective stems
                                                                            };

    static readonly Dictionary<string, List<string>> TwoCharacterMappings = new()
                                                                            {
                                                                               { "する", ["し", "さ", "すれ", "し", "せ"] },
                                                                               { "くる", ["き", "こ", "くれ", "き"] },
                                                                               { "いく", ["いき", "いか", "いけ", "いっ", "いこ"] },
                                                                               { "行く", ["行き", "行か", "行け", "行っ", "行こ"] },
                                                                               { "ます", ["まし", "ませ"] },
                                                                               { "いい", ["よく", "よけ", "よか", "よかっ"] }
                                                                            };

    static readonly List<string> MasuFormsByIndex = ["まし", "ませ", "まし", "まし"];

    static readonly HashSet<string> AruVerbs = ["なさる", "くださる", "おっしゃる", "ござる", "らっしゃる", "下さる", "為さる"];

    static readonly Dictionary<string, List<string>> AruMappings = new()
                                                                   {
                                                                      { "さる", ["さい", "さら", "され", "さっ"] },
                                                                      { "ざる", ["ざい", "ざら", "ざれ", "ざっ"] },
                                                                      { "ゃる", ["ゃい", "ゃら", "れば", "ゃっ"] }
                                                                   };

    static Conjugator()
    {
        GodanRuOrIchidanEndings = new List<string>(GodanRuEndings);
        GodanRuOrIchidanEndings.AddRange(IchidanEndings);
    }

    public static string ConstructRootVerbForPossiblyPotentialGodanVerbDictionaryForm(string potentialVerbForm)
    {
        var ending = potentialVerbForm.Substring(potentialVerbForm.Length - 2);
        return potentialVerbForm.Substring(0, potentialVerbForm.Length - 2) +
               GodanPotentialVerbEndingToDictionaryFormEndings[ending];
    }

    static bool IsAruVerb(string word) => AruVerbs.Any(word.EndsWith);

    public static List<string> GetWordStems(string word, bool isIchidanVerb = false, bool isGodan = false)
    {
        try
        {
            if (IsAruVerb(word))
            {
                var ending = word.Substring(word.Length - 2);
                return AruMappings[ending].Select(end => word.Substring(0, word.Length - 2) + end).ToList();
            }

            if (isIchidanVerb)
            {
                return IchidanEndings.Select(end => word.Substring(0, word.Length - 1) + end).ToList();
            }

            if (isGodan)
            {
                var lastChar = word.Substring(word.Length - 1);
                return OneCharacterMappings[lastChar].Select(end => word.Substring(0, word.Length - 1) + end).ToList();
            }

            var lastTwoChars = word.Length >= 2 ? word.Substring(word.Length - 2) : "";
            if (TwoCharacterMappings.ContainsKey(lastTwoChars))
            {
                return TwoCharacterMappings[lastTwoChars].Select(end => word.Substring(0, word.Length - 2) + end).ToList();
            }

            var lastOneChar = word.Substring(word.Length - 1);
            if (OneCharacterMappings.ContainsKey(lastOneChar))
            {
                if (lastOneChar == "る")
                {
                    return GodanRuOrIchidanEndings.Select(end => word.Substring(0, word.Length - 1) + end).ToList();
                }
                return OneCharacterMappings[lastOneChar].Select(end => word.Substring(0, word.Length - 1) + end).ToList();
            }
        }
        catch (KeyNotFoundException)
        {
            MyLog.Warning($"GetWordStems failed to handle {word}, returning empty list");
        }
        catch (System.ArgumentOutOfRangeException)
        {
            MyLog.Warning($"GetWordStems failed to handle {word}, returning empty list");
        }

        return [word];
    }

    static string GetStem(string word, int stemIndex, bool isIchidanVerb = false, bool isGodan = false)
    {
        try
        {
            if (IsAruVerb(word))
            {
                var ending = word.Substring(word.Length - 2);
                return word.Substring(0, word.Length - 2) + AruMappings[ending][stemIndex];
            }

            if (isIchidanVerb)
            {
                return word.Substring(0, word.Length - 1);
            }

            if (isGodan)
            {
                var lastChar = word.Substring(word.Length - 1);
                return word.Substring(0, word.Length - 1) + OneCharacterMappings[lastChar][stemIndex];
            }

            var lastTwoChars = word.Length >= 2 ? word.Substring(word.Length - 2) : "";
            if (lastTwoChars == "ます")
            {
                return word.Substring(0, word.Length - 2) + MasuFormsByIndex[stemIndex];
            }

            if (TwoCharacterMappings.ContainsKey(lastTwoChars))
            {
                return word.Substring(0, word.Length - 2) + TwoCharacterMappings[lastTwoChars][stemIndex];
            }

            var lastOneChar = word.Substring(word.Length - 1);
            if (OneCharacterMappings.ContainsKey(lastOneChar))
            {
                if (lastOneChar != "る")
                {
                    return word.Substring(0, word.Length - 1) + OneCharacterMappings[lastOneChar][stemIndex];
                }
                return word.Substring(0, word.Length - 1) + OneCharacterMappings[lastOneChar][stemIndex];
            }
        }
        catch (KeyNotFoundException)
        {
            MyLog.Warning($"GetStem failed to handle {word}, returning word as-is");
        }
        catch (System.ArgumentOutOfRangeException)
        {
            MyLog.Warning($"GetStem failed to handle {word}, returning word as-is");
        }

        return word;
    }

    public static List<string> GetVocabStems(VocabNote vocab) => GetWordStems(vocab.GetQuestion(), vocab.PartsOfSpeech.IsIchidan(), vocab.PartsOfSpeech.IsGodan());

    public static string GetIStem(string word, bool isIchidan = false, bool isGodan = false) => GetStem(word, IStemIndex, isIchidan, isGodan);

    public static string GetAStem(string word, bool isIchidan = false, bool isGodan = false) => GetStem(word, AStemIndex, isIchidan, isGodan);

    public static string GetEStem(string word, bool isIchidan = false, bool isGodan = false)
    {
        if (isIchidan)
        {
            return word.Substring(0, word.Length - 1) + "れ";
        }
        return GetStem(word, EStemIndex, isIchidan, isGodan);
    }

    public static string GetTeStem(string word, bool isIchidan = false, bool isGodan = false) => GetStem(word, TeStemIndex, isIchidan, isGodan);

    public static string GetIStemVocab(VocabNote vocab, string form = "")
    {
        var word = !string.IsNullOrEmpty(form) ? form : vocab.GetQuestion();
        return GetIStem(word, vocab.PartsOfSpeech.IsIchidan(), vocab.PartsOfSpeech.IsGodan());
    }

    public static string GetEStemVocab(VocabNote vocab, string form = "")
    {
        var word = !string.IsNullOrEmpty(form) ? form : vocab.GetQuestion();
        return GetEStem(word, vocab.PartsOfSpeech.IsIchidan(), vocab.PartsOfSpeech.IsGodan());
    }

    public static string GetAStemVocab(VocabNote vocab, string form = "")
    {
        var word = !string.IsNullOrEmpty(form) ? form : vocab.GetQuestion();
        return GetAStem(word, vocab.PartsOfSpeech.IsIchidan(), vocab.PartsOfSpeech.IsGodan());
    }

    public static string GetTeStemVocab(VocabNote vocab, string form = "")
    {
        var word = !string.IsNullOrEmpty(form) ? form : vocab.GetQuestion();
        return GetTeStem(word, vocab.PartsOfSpeech.IsIchidan(), vocab.PartsOfSpeech.IsGodan());
    }

    public static string GetImperative(string word, bool isIchidan = false, bool isGodan = false)
    {
        if (isGodan || (!isIchidan && !word.EndsWith("る")))
        {
            return GetEStem(word, isIchidan, isGodan);
        }

        return word.Substring(0, word.Length - 1) + "ろ";
    }
}
